Updated DatePartitionedQueryPlanner to succeed as long as one

ivakegg opened the pull request:
of the date ranges succeeds

ivakegg committed:
Updated DatePartitionedQueryPlanner to succeed as long as one
of the date ranges succeeds

ivakegg committed:
Updated to pass messages from the query planner through to the response
object

hlgp commented on warehouse/query-core/src/test/java/datawave/query/planner/DatePartitionedQueryPlannerTest.java:
```diff
@@ -878,6 +892,37 @@ public void testFieldIndexMinThresholdWithSomeNotMeeting() throws Exception {
+        givenEndDate("20130105");
+        givenPlan("(GENDER == 'male' || GENERE == 'male' || UUID == 'capone' || UUID == 'corleone')");
+
+        expectPlan(start("20130101"), end("20130103"), "GENERE == 'male' || UUID == 'capone' || UUID == 'corleone' || ((_Eval_ = true) && (GENDER == 'male'))");
```
In the "Hey Google" meeting, we talked about changing logic where a plan ends up being

EXECUTABLE || FULL TABLE SCAN

to proceed with the executable bit and no longer fail the whole thing.  Did we change our minds there?
Still on partial vacation brain, though, might be missing something....

ivakegg replied:
I think that is exactly what I did here.  As long as one of plans succeeds (i.e. is executable), then the query will run.  Did I fail to do that here?

hlgp replied:
Maybe I'm misreading assertQueryResults, but it looks like you have partialSuccessExpected = false, which reads as no partial success.  You also have fullTableScanRequired = true, which in the real world we do not allow.  Together it reads like this query will only succeed if full table scan is enabled, and I thought you were trying to demonstrate the opposite.

partialSuccessExpected-" This denotes that with full table scan disabled, success is expected anyway (at least one date range will succeed)."  I would expect from this description that your second argument should be true?  To show that at least one date range will succeed?  Is there something I'm missing in here?

ivakegg replied:
Ah, I see the problem here.  You are misreading the assertQueryResults.  The full table scan required is initially set to allow us to fully test the query plans.  If it is true, then we subsequently try it out with full table scan disabled to allow us to test whether we can partially be successful.  In this particular test case, I am testing that we fail because ALL of the date partitions fail.

hlgp replied:
Gotcha, thanks.

(Probably worth renaming the fullTableScanRequired to fullyTestQueryPlans and internally handle the full scan logic)

ivakegg committed:
Merge branch 'integration' into feature/heyGoogleSimple

ivakegg committed:
Updated test assertion method parameters per review comments

ivakegg committed:
Merge remote-tracking branch 'origin/integration' into feature/heyGoogleSimple

ivakegg committed:
Updated the planner to avoid value expansion until creating subplans

ivakegg committed:
Refactored the query planner to better organize the processing of the
plan, the reprocessing of the plan for date partitions, and the
starting of the plan stream.

ivakegg committed:
Working through refactoring of DefaultQueryPlanner for use with
DatePartitionedQueryPlanner

ivakegg committed:
Merge branch 'integration' into feature/heyGoogleSimple

ivakegg committed:
fix

ivakegg committed:
Test updates

ivakegg committed:
Merge branch 'integration' into feature/heyGoogleSimple

ivakegg committed:
Revert unused diffs

ivakegg committed:
Merge branch 'feature/heyGoogleSimple' of github.com:NationalSecurityAgency/datawave into feature/heyGoogleSimple

ivakegg committed:
Undo unused changes

ivakegg committed:
Trying to make diffs easier to read

ivakegg committed:
Upodated to separate expanding values for unfielded expansions vs
everthing else

apmoriarty commented on warehouse/query-core/src/main/java/datawave/query/config/ShardQueryConfiguration.java:
```diff
@@ -555,6 +556,12 @@ public class ShardQueryConfiguration extends GenericQueryConfiguration implement
+     * This property is used to avoid the pushdown/pullup portion of the planning. This is used by the DatePartitionedQueryPlanner as the pushdown/pullup is
+     * done by the reprocess planning method.
+     */
+    private boolean deferPushdownPullup = false;
```
Add `deferPushdownPullup` and `expandUnfieldedValues` to the equals and hashcode methods

ivakegg replied:
Will do.   Also adding tests to the ShardedQueryConfigurationTest to ensure everything is covered in equals and hashcode methods.

apmoriarty commented on warehouse/query-core/src/main/java/datawave/query/tables/ShardQueryLogic.java:
```diff
@@ -328,6 +329,7 @@ public String getPlan(AccumuloClient client, Query settings, Set<Authorizations>
                             + (this.getSettings() == null ? "empty" : this.getSettings().getId()) + ')');
         this.config.setExpandFields(expandFields);
         this.config.setExpandValues(expandValues);
+        this.config.setExpandUnfieldedValues(expandValues);
```
Add getter/setter for `expandUnfieldedValues`, I'm not sure if `deferPushdownPullup` requires a getter/setter here.

ivakegg replied:
Not sure we need it for expandUnfieldedValues, but I could see a case being made so I will add them.  Definitely not needed for deferPushdownPullup.

apmoriarty commented on warehouse/query-core/src/test/java/datawave/query/planner/DatePartitionedQueryPlannerTest.java:
```diff
@@ -622,18 +630,17 @@ public void testFieldIndexHolesPartiallyWithinDateRange() throws Exception {
         expectPlan(start("20130101"), end("20130102"), "UUID == 'capone' || UUID == 'corleone' || UUID == 'soprano'");
-        expectPlan(start("20130103"), end("20130104"),
-                        "((_Eval_ = true) && (UUID == 'capone')) || ((_Eval_ = true) && (UUID == 'corleone')) || ((_Eval_ = true) && (UUID == 'soprano'))");
+        expectPlan(start("20130103"), end("20130104"), "((_Eval_ = true) && (UUID =~ 'c.*')) || ((_Eval_ = true) && (UUID =~ 's.*'))");
```
Not necessary for this change set, consider in the future wrapping this in a method like `expectPartition(start, end, plan)`

ivakegg replied:
I kinda like it the way it is if you don't mind.

apmoriarty reviewed (changes requested):
Just the two changes to ShardQueryConfig and possibly ShardQueryLogic.

Query planner changes and test updates look good.

ivakegg committed:
Updated per review comments

ivakegg committed:
Merge branch 'integration' into feature/heyGoogleSimple

apmoriarty commented on warehouse/query-core/src/main/java/datawave/query/config/ShardQueryConfiguration.java:
```diff
@@ -3175,7 +3193,23 @@ public boolean equals(Object o) {
+                Objects.equals(getSummaryIterator(), that.getSummaryIterator()) &&
+                Objects.equals(getDisallowedRegexPatterns(), that.getDisallowedRegexPatterns()) &&
                 Objects.equals(getNoExpansionIfCurrentDateTypes(), that.getNoExpansionIfCurrentDateTypes()) &&
+                isDateIndexIterator() == that.isDateIndexIterator() &&
+                isUseDocumentScheduler() == that.isUseDocumentScheduler() &&
+                Objects.equals(getSummaryFieldName(), that.getSummaryFieldName()) &&
+                Objects.equals(getShardDateFormatter(), that.getShardDateFormatter()) &&
+                Objects.equals(getDocumentScannerConfig(), that.getDocumentScannerConfig()) &&
+                Objects.equals(getExcerptIterator(), that.getExcerptIterator()) &&
+                isUseQueryTreeScanHintRules() == that.isUseQueryTreeScanHintRules() &&
+                getMaxLinesToPrint() == that.getMaxLinesToPrint() &&
+                getMaxAnyFieldScanTimeMillis() == that.getMaxAnyFieldScanTimeMillis() &&
+                isDisableIteratorUniqueFields() == that.isDisableIteratorUniqueFields() &&
+                isUseShardedIndex() == that.isUseShardedIndex() &&
```
good catches

ivakegg committed:
Merge branch 'integration' into feature/heyGoogleSimple

drewfarris commented on warehouse/query-core/src/main/java/datawave/query/planner/DatePartitionedQueryPlanner.java:
```diff
@@ -302,66 +304,87 @@ public CloseableIterable<QueryData> process(GenericQueryConfiguration genericCon
         }
 
         // Let's do the planning with the delegate planner first to ensure we have a final date range
         // and appropriately expanded unfielded terms etc.
+        boolean generatePlanOnly = planningConfig.isGeneratePlanOnly();
+        planningConfig.setGeneratePlanOnly(true);
+        boolean expandValues = planningConfig.isExpandValues();
+        // we do NOT want to expand any values yet as they may not be dependable
+        // note we are expanding unfielded values (different flag)
+        planningConfig.setExpandValues(false);
+        planningConfig.setDeferPushdownPullup(true);
+
```
Minor nit: if we preserve the original `generatePlanOnly` and `expandValues` value so we can add them back to the config later, why don't we do the same for `deferPushdownPullup`?

ivakegg replied:
That is not a value that we change via the query logic, so I did not see a need to do that.

ivakegg replied:
However I will do it to be safe...

ivakegg committed:
Updated per review comment

ivakegg committed:
Merge branch 'integration' into feature/heyGoogleSimple
